aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/transfer/internal/CompleteMultipartDownload.java

```diff
@@ -41,13 +43,23 @@ public CompleteMultipartDownload(List<Future<Long>> files, File destinationFile,
 
     @Override
     public File call() throws Exception {
+        int index = 0;
         try {
-            for (Future<Long> file : partFiles) {
-                long filePosition = file.get();
+            for (; index < partFiles.size(); index++) {
+                long filePosition = partFiles.get(index).get();
                 download.updatePersistableTransfer(currentPartNumber++, filePosition);
             }
-
             download.setState(Transfer.TransferState.Completed);
+        } catch (ExecutionException e) {
+            // if any part fails, we cancel remaining part downloads and notify clients of the failure.
+            for (; index < partFiles.size(); index++){
```

Comment:
Based on this implementation, all the uploads next to the exception-upload will be canceled (including the upload which got exception). So what about the uploads happened before to this exception (which are still pending at this point)?

Reply:
Since each individual piece is a future to an executor, which may or may not be a single executor, you can fail out of order. For example, you can have this failure 

Blocks => D = downloaded, f = failure, p = pending

**Ex1:**
file = { D, D, D, f, p , p, D}

But you cannot have this failure:
**Ex2:**
file = {D, p, D, f, _, _, _, _}

or this failure:
**Ex3:**
file = {D, f, D, f, _, _, _, _}


In "Ex1" your file wrote 3 blocks, has 2 blocks pending and 1 block written at the end that's good. You'll lose that block, but you would anyway if you paused the download at that point.

"Ex2" and "Ex3" cannot exist because CompletableDownload forces assembly in serial order.

The PersitableDownload feature, and the CompletableDownload are both based on serial order. That's a constraint within the existing implementation. 

When you say "uploads" it's a little bit ambiguous to me if you are referring to a "part upload" or a "file upload". To try to make this clear, the CompletableDownload is a container for a single file, and each future is a part of that file. Which means (as explained above), on failure, you'll get the a continuous sequence of completed parts from the "start" to block N. Which will allow the caller to, if they so chose, get a handle to the PersitableDownload at which they can resume at a later time.

Hope it helps.